Empire/EmpireManager.cpp

```diff
@@ -177,91 +207,147 @@ void EmpireManager::SetDiplomaticMessage(const DiplomaticMessage& message) {
     int empire2 = message.RecipientEmpireID();
     const DiplomaticMessage& initial_message = GetDiplomaticMessage(empire1, empire2);
     if (message != initial_message) {
-        m_diplomatic_messages[DiploKey(empire1, empire2)] = message;
+        m_diplomatic_messages[{empire1, empire2}] = message;
         DiplomaticMessageChangedSignal(empire1, empire2);
     }
 }
 
-void EmpireManager::RemoveDiplomaticMessage(int empire1, int empire2) {
-    std::pair<int, int> key = DiploKey(empire1, empire2);
-    std::map<std::pair<int, int>, DiplomaticMessage>::iterator it = m_diplomatic_messages.find(key);
+void EmpireManager::RemoveDiplomaticMessage(int sender_id, int recipient_id) {
+    auto it = m_diplomatic_messages.find({sender_id, recipient_id});
     if (it != m_diplomatic_messages.end()) {
-        m_diplomatic_messages[key] = DiplomaticMessage(empire1, empire2, DiplomaticMessage::INVALID_DIPLOMATIC_MESSAGE_TYPE);
-        DiplomaticMessageChangedSignal(empire1, empire2);
+        m_diplomatic_messages[{sender_id, recipient_id}] =
+            DiplomaticMessage(sender_id, recipient_id, DiplomaticMessage::INVALID_DIPLOMATIC_MESSAGE_TYPE);
+        DiplomaticMessageChangedSignal(sender_id, recipient_id);
         return;
     }
-    ErrorLogger() << "Was no diplomatic message entry between empires " << empire1 << " and " << empire2;
-    m_diplomatic_messages[key] = DiplomaticMessage(empire1, empire2, DiplomaticMessage::INVALID_DIPLOMATIC_MESSAGE_TYPE);
+    ErrorLogger() << "Was no diplomatic message entry between empires " << sender_id << " and " << recipient_id;
+    m_diplomatic_messages[{sender_id, recipient_id}] =
+        DiplomaticMessage(sender_id, recipient_id, DiplomaticMessage::INVALID_DIPLOMATIC_MESSAGE_TYPE);
 }
 
 void EmpireManager::HandleDiplomaticMessage(const DiplomaticMessage& message) {
+    DebugLogger() << "HandleDiplomaticMessage start message: " << message.Dump();
+    DebugLogger() << this->DumpDiplomacy();
+
     int sender_empire_id = message.SenderEmpireID();
     int recipient_empire_id = message.RecipientEmpireID();
-    //std::pair<int, int> key = DiploKey(sender_empire_id, recipient_empire_id);
+
     DiplomaticStatus diplo_status = GetDiplomaticStatus(sender_empire_id, recipient_empire_id);
-    bool message_already_available = DiplomaticMessageAvailable(sender_empire_id, recipient_empire_id);
-    const DiplomaticMessage& existing_message = GetDiplomaticMessage(sender_empire_id, recipient_empire_id);
+    bool message_from_recipient_to_sender_available = DiplomaticMessageAvailable(recipient_empire_id, sender_empire_id); 
+    const DiplomaticMessage& existing_message_from_recipient_to_sender = GetDiplomaticMessage(recipient_empire_id, sender_empire_id);
+    //bool message_from_sender_to_recipient_available = DiplomaticMessageAvailable(sender_empire_id, recipient_empire_id);
```

Comment:
I don't know how to clarify... it's checking if there is already another message waiting from the recipient of the current message to the sender of the current message. ie. when handing a message from A to B, it checks if there was already a message from B to A sent.